src/digraph.rs

```diff
@@ -971,6 +973,103 @@ impl PyDiGraph {
         }
         Err(NoSuitableNeighbors::py_err("No suitable neighbor"))
     }
+
+    /// Generate a dot file from the graph
+    ///
+    /// :param node_attr: A callable that will take in a node data object
+    ///     and return a dictionary of attributes to be associated with the
+    ///     node in the dot file. The key and value of this dictionary **must**
+    ///     be a string. If they're not strings retworkx will raise TypeError
+    ///     (unfortunately without an error message because of current
+    ///     limitations in the PyO3 type checking)
+    /// :param edge_attr: A callable that will take in an edge data object
+    ///     and return a dictionary of attributes to be associated with the
+    ///     node in the dot file. The key and value of this dictionary **must**
+    ///     be a string. If they're not strings retworkx will raise TypeError
+    ///     (unfortunately without an error message because of current
+    ///     limitations in the PyO3 type checking)
+    /// :param dict graph_attr: An optional dictionary that specifies any graph
+    ///     attributes for the output dot file. The key and value of this
+    ///     dictionary **must** be a string. If they're not strings retworkx
+    ///     will raise TypeError (unfortunately without an error message
+    ///     because of current limitations in the PyO3 type checking)
+    /// :param str filename: An optional path to write the dot file to
+    ///     if specified there is no return from the function
+    ///
+    /// :returns: A string with the dot file contents if filename is not
+    ///     specified.
+    /// :rtype: str
+    ///
+    /// Using this method enables you to leverage graphviz to visualize a
+    /// :class:`retworkx.PyDiGraph` object. For example:
+    ///
+    /// .. jupyter-execute::
+    ///
+    ///   import os
+    ///   import tempfile
+    ///
+    ///   import pydot
+    ///   from PIL import Image
+    ///
+    ///   import retworkx
+    ///
+    ///   graph = retworkx.directed_gnp_random_graph(15, .25)
+    ///   dot_str = graph.to_dot(
+    ///       lambda node: dict(
+    ///           color='black', fillcolor='lightblue', style='filled'),
+    ///       lambda edge: {})
+    ///   dot = pydot.graph_from_dot_data(dot_str.decode('utf8'))[0]
+    ///
+    ///   with tempfile.TemporaryDirectory() as tmpdirname:
+    ///       tmp_path = os.path.join(tmpdirname, 'dag.png')
+    ///       dot.write_png(tmp_path)
+    ///       image = Image.open(tmp_path)
+    ///       os.remove(tmp_path)
+    ///   image
+    ///
+    #[text_signature = "(node_attr, edge_attr, /, graph_attr=None, filename=None)"]
```

Comment:
What does this line do?

Reply:
This line is a macro that defines the text signature for python function over the c api. It basically lets python's `inspect` and other tools that need a signature know the parameters on the function via a string. https://pyo3.rs/v0.11.1/function.html#making-the-function-signature-available-to-python